Sources/NIO/Codec.swift

```diff
@@ -674,3 +674,118 @@ extension MessageToByteHandler {
         }
     }
 }
+
+enum ByteToMessageDecoderVerificationError<OutputType>: Error {
+    case wrongDecodingResult(actual: OutputType, expected: OutputType)
+    case unexpectedDecodingResult(OutputType)
+    case leftOversOnDeconstructingChannel(inbound: [NIOAny], outbound: [NIOAny], pendingOutbound: [NIOAny])
+}
+
+extension ByteToMessageHandler {
+    /// Verifies a `ByteToMessageDecoder` by performing a number of tests.
+    ///
+    /// This method is mostly useful in unit tests for `ByteToMessageDecoder`s. It feeds the inputs from
+    /// `inputOutputPairs` into the decoder in various ways and expects the decoder to produce the outputs from
+    /// `inputOutputPairs`.
+    ///
+    /// The verification is performing various tests, for example:
+    ///
+    ///  - drip feeding the bytes, one by one
+    ///  - sending many messages in one `ByteBuffer`
+    ///  - sending each complete message in one `ByteBuffer`
+    ///
+    /// For `ExampleDecoder` that produces `ExampleDecoderOutput`s you would use this method the following way:
+    ///
+    ///     var exampleInput1 = channel.allocator.buffer(capacity: 16)
+    ///     exampleInput1.writeString("example-in1")
+    ///     var exampleInput2 = channel.allocator.buffer(capacity: 16)
+    ///     exampleInput2.writeString("example-in2")
+    ///     let expectedInOuts = [(exampleInput1, ExampleDecoderOutput("1")),
+    ///                           (exampleInput2, ExampleDecoderOutput("2"))
+    ///                          ]
+    ///     XCTAssertNoThrow(try ByteToMessageHandler<ExampleDecoder>.verifyDecoder(inputOutputPairs: expectedInOuts,
+    ///                                                                             decoderFactory: { ExampleDecoder() }))
+    public static func verifyDecoder<Decoder: ByteToMessageDecoder>(inputOutputPairs: [(ByteBuffer, [Decoder.InboundOut])],
```

Comment:
it's a bit odd to have this in the "normal" code-base... Is there a way to have "test-only" code, so users would explicit import it in test-code only. Maybe the only way would be a new "product"...

Reply:
> it's a bit odd to have this in the "normal" code-base... Is there a way to have "test-only" code

No, unfortunately not.

>  Maybe the only way would be a new "product"...

We could do that. @Lukasa ?

Reply:
I think a new product would probably be sensible.